File: skills/rango-integration/examples/session-memory.ts
// Session Memory Example — TypeScript
// Demonstrates persisting user sessions with governance metadata

import { RangoClient } from '@rango/core';
import { v7 as uuidv7 } from 'uuid';

function main() {
    // Open workspace
    const client = new RangoClient('./session-memory-example.rango', 'session-example');
    
    // Insert session with governance metadata
    const sessionId = uuidv7();
    const sessionDoc = JSON.stringify({
        user_id: 'alice',
        session_token: sessionId,
        created_at: new Date().toISOString(),
        tenant_id: 'org-123',              // Governance: tenant scope
        lineage: `session:${sessionId}`,   // Governance: provenance
        trust_score: 1.0,                 // Governance: verified internal
        verified: true,
        expires_at: new Date(Date.now() + 3600000).toISOString(),
    });
    
    const docId = client.insertOne('sessions', sessionDoc);
    console.log(`Session created: ${docId}`);
    
    // Retrieve session
    const doc = client.findOne('sessions', docId);
    if (doc) {
        console.log(`Found session: ${doc}`);
    }
}

main();

File: skills/rango-integration/examples/session-memory.rs
// Session Memory Example — Rust
// Demonstrates persisting user sessions with governance metadata

use std::sync::Arc;
use bson::{doc, Document};
use rango_sdk::RangoClient;
use rango_storage::{DegradingStorage, RedbStorage};
use rango_oplog::FileOplog;
use rango_types::*;
use uuid::Uuid;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Open workspace
    let path = std::path::Path::new("./session-memory-example.rango");
    std::fs::create_dir_all(path)?;
    
    let storage_path = path.join("data.redb");
    let inner = RedbStorage::open(&storage_path)?;
    let storage = Arc::new(
        DegradingStorage::with_default_threshold(inner, &storage_path)?
    );
    
    let oplog = Arc::new(FileOplog::new(path.join("oplog.bin"))?);
    let client = RangoClient::open(storage, oplog, "session-example")?;
    
    // Insert session with governance metadata
    let session_id = Uuid::new_v7().to_string();
    let session_doc = doc! {
        "user_id": "alice",
        "session_token": session_id.clone(),
        "created_at": bson::DateTime::now(),
        "tenant_id": "org-123",          // Governance: tenant scope
        "lineage": format!("session:{}", session_id), // Governance: provenance
        "trust_score": 1.0,              // Governance: verified internal
        "verified": true,
        "expires_at": bson::DateTime::now().timestamp_millis() + 3600000i64,
    };
    
    let sessions = client.collection("sessions");
    let doc_id = sessions.insert_one(session_doc)?;
    
    println!("Session created: {}", doc_id);
    
    // Retrieve session
    if let Some(doc) = sessions.find_one(&doc_id)? {
        println!("Found session: {:?}", doc.data);
    }
    
    Ok(())
}
